src/index.ts

```diff
@@ -43,93 +132,49 @@ export class I18N {
-            );
-
-            return key;
-        }
+        let text: string | undefined;
+        let fallbackText: string | undefined;
+        let details: TranslationData['details'];
 
-        if (Object.keys(keyset).length === 0) {
-            this.warn(
-                'Keyset is empty.',
-                keysetName,
-            );
+        ({text, fallbackText, details} = this.getTranslationData({keysetName, key, params, lang: this.lang}));
 
-            return key;
+        if (details) {
+            const message = mapErrorCodeToMessage({
+                code: details.code,
+                lang: this.lang,
+                fallbackLang: this.fallbackLang === this.lang ? undefined : this.fallbackLang,
+            });
+            this.warn(message, details.keysetName, details.key);
         }
 
-        const keyValue = keyset && keyset[key];
-        let result: string;
-
-        if (typeof keyValue === 'undefined') {
-            this.warn(
-                'Missing key.',
+        if (text === undefined && this.fallbackLang && this.fallbackLang !== this.lang) {
+            ({text, fallbackText, details} = this.getTranslationData({
                 keysetName,
                 key,
-            );
-
-            return key;
-        }
-
-        if (Array.isArray(keyValue)) {
-            if (keyValue.length < 3) {
-                this.warn(
-                    'Missing required plurals',
-                    keysetName,
-                    key,
-                );
-
-                return key;
-            }
-
-            const count = Number(params?.count);
-
-            if (Number.isNaN(count)) {
-                this.warn(
-                    'Missing params.count for key.',
-                    keysetName,
-                    key,
-                );
-
-                return key;
+                params,
+                lang: this.fallbackLang,
+            }));
+
+            if (details) {
+                const message = mapErrorCodeToMessage({
+                    code: details.code,
+                    lang: this.fallbackLang,
+                });
+                this.warn(message, details.keysetName, details.key);
             }
-
-            const pluralizer = this.getLanguagePluralizer(this.lang);
-            result = keyValue[pluralizer(count, PluralForm)] || keyValue[PluralForm.Many];
-
-            if (keyValue[PluralForm.None] === undefined) {
-                this.warn('Missing key for 0', keysetName, key);
-            }
-        } else {
-            result = keyValue;
-        }
-
-        if (params) {
-            result = replaceParams(result, params);
         }
 
-        return result;
+        // this.getTranslationData method necessarily returns either a `text` or a `fallbackText` value.
+        // Both of these fields are described in the `TranslationData` type as optional so as not to complicate the code.
+        return (text ?? fallbackText) as string;
```

Comment:
Let's remove `fallbackText` value from  the `TranslationData`
```suggestion
        return text ?? key;
```